subsys/net/ip/tcp.c

```diff
@@ -748,21 +749,57 @@ static size_t tcp_check_pending_data(struct tcp *conn, struct net_pkt *pkt,
 
 	if (CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT &&
 	    !net_pkt_is_empty(conn->queue_recv_data)) {
+		/* Some potentential cases:
+		 * Note: MI = MAX_INT
+		 * Packet | Queued| End off   | Gap size | Required handling
+		 * Seq|Len|Seq|Len|           |          |
+		 *  3 | 3 | 6 | 4 | 3+3-6=  0 | 6-3-3=0  | Append
+		 *  3 | 4 | 6 | 4 | 3+4-6 = 1 | 6-3-4=-1 | Append, pull from queue
+		 *  3 | 7 | 6 | 4 | 3+7-6 = 4 | 6-3-7=-4 | Drop queued data
+		 *  3 | 8 | 6 | 4 | 3+8-6 = 5 | 6-3-8=-5 | Drop queued data
+		 *  6 | 5 | 6 | 4 | 6+5-6 = 5 | 6-6-5=-5 | Drop queued data
+		 *  6 | 4 | 6 | 4 | 6+4-6 = 4 | 6-6-4=-4 | Drop queued data / packet
+		 * 10 | 2 | 6 | 4 | 10+2-6= 6 | 6-10-2=-6| Should not happen, dropping queue
+		 *  7 | 4 | 6 | 4 | 7+4-6 = 5 | 6-7-4=-5 | Should not happen, dropping queue
+		 * 11 | 2 | 6 | 4 | 11+2-6= 7 | 6-11-2=-7| Should not happen, dropping queue
+		 *  2 | 3 | 6 | 4 | 2+3-6= MI | 6-2-3=1  | Keep queued data
+		 */
 		struct tcphdr *th = th_get(pkt);
 		uint32_t expected_seq = th_seq(th) + len;
 		uint32_t pending_seq;
+		uint32_t gap_size;
```

Comment:
Why can't the `gap_size` simply be an `int32_t` variable? It seems a bit of an overhead to me to have a separate pointer variable just to do the casting.

Reply:
Could be, does that work in all cases correctly with overflowing?

Reply:
It is no different than using a pointer, if the MSB is set the value will be interpreted as a negative. You may need to add an explicit cast to avoid warning (but we know what we're doing here):
`gap_size = (int32_t)(pending_seq - th_seq(th) - len);`

Reply:
Thanks for the input, I applied your suggestion, with the added explicit cast of len to uint32_t